apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1360,8 +1395,221 @@ private void StartPush()
+
+                    bool hasAcksOrNacks = ackOrNackIds.Any();
+                    bool hasExtends = extendIds.Any();
+                    if (!hasAcksOrNacks && !hasExtends)
+                    {
+                        return;
+                    }
+
+                    var idsToRetry = hasAcksOrNacks ? ackOrNackIds : extendIds.Select(j => j.Id);
+                    var retryableErrorIds = new ConcurrentBag<string>();
+                    // Not all temporaryFailures may need to be retried.
+                    // See if the id exists in the _retryableIds dictionary and only then retry.
+                    foreach (var id in idsToRetry.Where(j => _retryableIds.ContainsKey(j)))
+                    {
+                        retryableErrorIds.Add(id);
+                    }
+
+                    if (retryableErrorIds.Count == 0)
+                    {
+                        // Nothing to retry.
+                        return;
+                    }
+
+                    // All ids in retryableErrorIds will have the same backoff.
+                    // Get the retryInfo for the first retryableErrorId from the _retryableIds dictionary.
+                    if (!_retryableIds.TryGetValue(retryableErrorIds.First(), out var retryInfo))
+                    {
+                        // Nothing to retry.
+                        return;
+                    }
+
+                    var backoff = s_ackBackoff.NextBackoff(retryInfo.CurrentBackoff ?? TimeSpan.Zero);
+                    // We should retry only for specified timeout.
+                    // If current UTC time is greater than first UTC time of failure + timeout, we should not retry.
+                    if (DateTime.UtcNow > retryInfo.FirstTimeOfFailureInUtc + TimeSpan.FromSeconds(timeoutInSeconds) - backoff)
+                    {
+                        // We are past the retry timeout.
+                        return;
+                    }
+
+                    // If we have ids to retry and we are not past the specified timeout, then retry.
+                    // Update backoff of retryableErrorIds in _retryableIds dictionary.
+                    UpdateAllRetryableErrorBackoff(retryableErrorIds, backoff);
+                    // Retry errors that are still temporary errors. They will be retried after backOff.
+                    Task delayTask = _scheduler.Delay(backoff, _softStopCts.Token);
+                    // TODO: Check, if the below expression can be simplified or optimized.
+                    Add(delayTask, new NextAction(false, hasAcksOrNacks
+                    ? () => { ackActionToRetry(retryableErrorIds); StartPush(); }
+                    : () => { extendActionToRetry(extendIds.Where(j => retryableErrorIds.Contains(j.Id))); StartPush(); }));
+                }
+
+                // Updates the backoff of the retryableErrorIds in _retryableIds dictionary.
```

Comment:
Presumably this is safe to do in the "try-get then update" because for any given ID, we won't be running this code while there's an in-flight request that could complete (and run it separately)?

Reply:
We are only processing the response for the given ackIds/nackIds/extendIds and know their latest status before scheduling them again with a backoff, so we should be good.